### PR TITLE
Add file num and size metrics for each level file

### PR DESCRIPTION
**Pull request by THUMarkLau**
This PR add metrics for file count and size in each compaction level, which may help us to tune the system online.

image

### CODE REVIEW

**Comment by THUMarkLau** [comment]
This PR is set to draft until I fix the code smell next week.

File: server/src/main/java/org/apache/iotdb/db/engine/TsFileMetricManager.java [hunk 1]
```diff
@@ -50,30 +71,154 @@ public class TsFileMetricManager {
...
+    if (metricService != null) {
+      metricService
+          .getOrCreateGauge(
+              FILE_LEVEL_COUNT,
+              MetricLevel.CORE,
+              Tag.TYPE.toString(),
+              seq ? SEQUENCE : UNSEQUENCE,
+              LEVEL,
+              String.valueOf(level))
+          .set(count);
```
Comment (SpriCoder) [inline server/src/main/java/org/apache/iotdb/db/engine/TsFileMetricManager.java:137]: Consider avoiding the acquire of gauge from metric service. Instead, we can create a metric set to manage all related metrics, including creation, update and remove, more details: https://github.com/apache/iotdb/pull/9901

**Review by SpriCoder — CHANGES_REQUESTED** [review]
PTAL~

File: server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java [hunk 2]
```diff
@@ -65,6 +65,7 @@ public void bindTo(AbstractMetricService metricService) {
   }
 
   private void bindTsFileMetrics(AbstractMetricService metricService) {
+    TS_FILE_METRIC_MANAGER.setMetricService(metricService);
```
Comment (SpriCoder) [inline server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java:68]: A little bit strange. We use TS_FILE_METRIC_MANAGER to manage metrics and add metric service into TS_FILE_METRIC_MANAGER. From my perspective, I think we can use TS_FILE_METRIC_MANAGER to manage related value, and use Metric Set to manage metric(get value from TS_FILE_METRIC_MANAGER)

**Review by SpriCoder — CHANGES_REQUESTED** [review]
PTAL~

File: server/src/main/java/org/apache/iotdb/db/engine/TsFileMetricManager.java [hunk 3]
```diff
@@ -50,30 +73,155 @@ public class TsFileMetricManager {
...
+  }
+
+  public void addFile(long size, boolean seq, String name) {
+    updateGlobalCountAndSize(size, 1, seq);
+    try {
+      TsFileNameGenerator.TsFileName tsFileName = TsFileNameGenerator.getTsFileName(name);
+      int level = tsFileName.getInnerCompactionCnt();
+      updateLevelCountAndSize(size, 1, seq, level);
+    } catch (IOException e) {
+      log.error("Unexpected error occurred when getting tsfile name", e);
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/engine/TsFileMetricManager.java:100]: use warn?

File: server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java [hunk 4]
```diff
@@ -277,8 +326,234 @@ private long getOpenFileHandlersNumber() {
...
+                    FILE_LEVEL_SIZE,
+                    MetricLevel.CORE,
+                    Tag.TYPE.toString(),
+                    orderStr,
+                    LEVEL,
+                    String.valueOf(level)))
+        .set(size);
+  }
+
+  public void deleteFile(List<Long> sizeList, boolean seq, int num, List<String> names) {
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java:437]: remove num

File: server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java [hunk 5]
```diff
@@ -38,26 +43,70 @@
...
+
+  private final AtomicInteger modFileNum = new AtomicInteger(0);
+
+  private final AtomicLong modFileSize = new AtomicLong(0);
+  private final Map<Integer, Integer> seqLevelTsFileCountMap = new ConcurrentHashMap<>();
+  private final Map<Integer, Integer> unseqLevelTsFileCountMap = new ConcurrentHashMap<>();
+  private final Map<Integer, Long> seqLevelTsFileSizeMap = new ConcurrentHashMap<>();
+  private final Map<Integer, Long> unseqLevelTsFileSizeMap = new ConcurrentHashMap<>();
+  private long lastUpdateTime = 0;
+  private static final long UPDATE_INTERVAL = 10_000L;
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/service/metrics/FileMetrics.java:81]: Put that value in a more basic class?

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/execute/task/CrossSpaceCompactionTask.java [hunk 6]
```diff
@@ -349,16 +353,16 @@ public void resetCompactionCandidateStatusForAllSourceFiles() {
     selectedUnsequenceFiles.forEach(x -> x.setStatus(TsFileResourceStatus.NORMAL));
   }
 
-  private long deleteOldFiles(List<TsFileResource> tsFileResourceList) throws IOException {
-    long totalSize = 0;
+  private List<Long> deleteOldFiles(List<TsFileResource> tsFileResourceList) {
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/execute/task/CrossSpaceCompactionTask.java:356]: use long[]

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/execute/task/CrossSpaceCompactionTask.java [hunk 7]
```diff
@@ -191,28 +191,36 @@ public boolean doCompaction() {
...
+            FileMetrics.getInstance().decreaseModFileNum(1);
+            FileMetrics.getInstance()
                 .decreaseModFileSize(unsequenceResource.getModFile().getSize());
           }
         }
 
-        long sequenceFileSize = deleteOldFiles(selectedSequenceFiles);
-        long unsequenceFileSize = deleteOldFiles(selectedUnsequenceFiles);
+        List<Long> sequenceFileSize = deleteOldFiles(selectedSequenceFiles);
+        List<String> fileNames = new ArrayList<>();
```
Comment (OneSizeFitsQuorum) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/execute/task/CrossSpaceCompactionTask.java:208]: Avoid extended copying of arrays

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**